src/mate/academy/hashMap/ReferenceMyHashMap.java

```diff
@@ -0,0 +1,246 @@
+package mate.academy.hashMap;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+public class ReferenceMyHashMap<K, V> implements MyHashMap<K, V> {
+
+    private Node<K, V>[] hashTable;
+    private int size = 0;
+    private float threshold;
+
+    public ReferenceMyHashMap() {
+        hashTable = new Node[16];
+        threshold = hashTable.length * 0.75F;
+    }
+
+    @Override
+    public boolean put(final K key, final V value) {
+        if (size + 1 >= threshold) {
+            threshold *= 2;
+            arrayDoubling();
+        }
+
+        Node<K, V> newNode = new Node<>(key, value);
+        int index = hash(key);
+
+        if (hashTable[index] == null) {
+            return simpleAdd(index, newNode);
+        }
+        //get node of this index
+        List<Node<K, V>> nodeList = hashTable[index].getNodes();
+        //run of nodeList collection
+        for (Node<K, V> node : nodeList) {
+            //if the node is exist, we should override him
+            if (keyExistButValueNew(node, newNode, value) ||
+                    collisionProcessing(node, newNode, nodeList)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void arrayDoubling() {
+        Node<K, V>[] oldHashTable = hashTable;
+        hashTable = new Node[oldHashTable.length * 2];
+        size = 0;
+        for (Node<K, V> node : oldHashTable) {
+            if (node != null) {
+                for (Node<K, V> n : node.getNodes()) {
+                    put(n.key, n.value);
+                }
+            }
+        }
+    }
+
+    private boolean collisionProcessing(
+            final Node<K, V> nodeFromList,
+            final Node<K, V> newNode,
+            final List<Node<K, V>> nodes) {
+        /*if hashcode of new node = hash code node from list,
+        != key and value we should add new node + increment the size
+        */
+        if (newNode.hashCode() == nodeFromList.hashCode() &&
+                !Objects.equals(newNode.key, nodeFromList.key) &&
+                !Objects.equals(newNode.value, nodeFromList.value)
+        ) {
+            nodes.add(newNode);
+            size++;
+            return true;
+        }
+        return false;
+    }
+
+    private boolean keyExistButValueNew(
+            final Node<K, V> nodeFromList,
+            final Node<K, V> newNode,
+            final V value) {
+        //if the key is coincide and the value not coincide...
```

Comment:
code should be understandable enough without comments, otherwise something is wrong which your code